nomad/leader.go

```diff
@@ -57,11 +57,26 @@ var defaultSchedulerConfig = &structs.SchedulerConfiguration{
 // as the leader in the Raft cluster. There is some work the leader is
 // expected to do, so we must react to changes
 func (s *Server) monitorLeadership() {
+	// We use the notify channel we configured Raft with, NOT Raft's
+	// leaderCh, which is only notified best-effort. Doing this ensures
+	// that we get all notifications in order, which is required for
+	// cleanup and to ensure we never run multiple leader loops.
+	leaderCh := s.leaderCh
+
 	var weAreLeaderCh chan struct{}
 	var leaderLoop sync.WaitGroup
 	for {
 		select {
-		case isLeader := <-s.leaderCh:
+		case isLeader := <-leaderCh:
+			// suppress leadership flapping
+			isLeader, suppressed := suppressLeadershipFlaps(isLeader, leaderCh)
+
+			// if gained and lost leadership immediately, move on without emitting error
+			if suppressed && !isLeader && weAreLeaderCh == nil {
+				s.logger.Info("cluster leadership acquired but lost immediately")
```

Comment:
Does flapping indicate something an operator should be aware of such as excessive cpu load or network issues? If so we might want to document that in at least a code comment as I'm worried this log message may flummox the first users who hit it. If it does hint at some degenerate state, users who see this might already be stressed and confused, so if there's any more hints or context we can offer we should.

Reply:
To add to this, I've often found as an operator that `info` log level is indecisive. If it's something I need to pay attention to and take action on, it should probably be at `warn`. Otherwise it's telling me "hey this is something you want to know about" without telling me what to do with that information.

(My rule of thumb has often been that `info` is for information at startup like "I'm listening on port 1234 and running with plugins X and Y", and then not used after that, but I don't think we have firm guidelines anywhere on Nomad.)

Reply:
Hm, are this eliminate `leaderCh` overflow? I think that is no. For example this is not remove situation when leadership lost, and you just increase  `leaderCh` buffer size which is not enough imho in all cases